### ISSUE TITLE
Statsmodels prints uninformative message

### ISSUE THREAD

**Issue by eXcuvator** [statsmodels/statsmodels#4118]
When running a major file of code I kept getting

> not in depth 1

and similar messages. I had no clue how to analyze this (in my larger system), and assumed it was an error message thrown by multiprocessing (something about workers that were inaccessible or thelike). 

After a text search in my whole site-packages, I found the culprit:

[In]( https://github.com/statsmodels/statsmodels/blob/master/statsmodels/base/data.py#L107) , it just says

```
            for depth in [2, 3, 1, 0, 4]:  # sequence is a guess where to likely find it
                try:
                    _, design = dmatrices(d['formula'], data, eval_env=depth,
                                          return_type='dataframe')
                    break
                except (NameError, PatsyError) as e:
                    print('not in depth %d' % depth)
                    exc.append(e)   # why do I need a reference from outside except block
                    pass
```

This error handling is confusing at the least.

**Comment by josef-pkt** [comment]
Thanks for the report,
I just see also that it uses a print and not a warnings.warn.

Can you provide an example or describe the usecase? Is this for pickling for multiprocessing?
Maybe there is a way to make this more robust

The main problem is that this is a hackish workaround that formulas cannot be pickled.

The specific print was mainly for debugging and I didn't expect that users will run into it, also the actual error message created by patsy should be in `e` 

Would it be enough to mention unpickling in the message?
e.g. 
`msg = 'error in unpickling formula and recreating dmatrices using depth  %d' % depth`

**Comment by eXcuvator** [comment]
Yes, the use case is to send an `smf.ols().fit()` object into multiprocessing. For that it has to be pickled.

```
    import statsmodels.formula.api as smf
    from pathos.multiprocessing import ProcessingPool

    predictor = smf.ols(foo, data=df).fit()
    pool = ProcessingPool(4)
    xs = pool.map(myFunc,
                     [predictor] * 10)
```

Unpickling in the message would be helpful, but also a way to ignore them. If it's a warnings.warn, I suppose I could turn that off via warnings.filterwarnings. 

Finally, I still don't know what "Error in ..." actually means. Is the design matrix potentially corrupt?

**Comment by josef-pkt** [comment]
> Finally, I still don't know what "Error in ..." actually means. Is the design matrix potentially corrupt?

The recreated `self.design_info` could be incorrect. AFAIR, we keep the design matrix separate if it hasn't been `remove_data`ed. But patsy's formula `design_info` can currently not be pickled so our workaround is to recreate it. The main potential issue is whether transformations that are in the formula are compatible or identical to the one in the original model. We don't have much control over this in statsmodels because it's between patsy and the user environment.
Our examples and test cases were all rather simple where the environment didn't change except for maybe the depth, i.e. whether defined in a local scope or a surrounding scope.
(I guess the main problem is that patsy uses `eval` to allow arbitrary user functions in the formula, which makes it difficult to limit to what has to be available, or what we can check or recreate.)

The `design_info` is used in a few places for post processing, e.g. the `transform` in predict.
